packages/@aws-cdk/aws-cloudwatch/lib/metric.ts

```diff
@@ -249,12 +264,14 @@ export class Metric implements IMetric {
       // For these we're not going to do deep equality, misses some opportunity for optimization
       // but that's okay.
       && (props.dimensions === undefined)
+      && (props.dimensionsMap === undefined)
       && (props.period === undefined || props.period.toSeconds() === this.period.toSeconds())) {
       return this;
     }
 
     return new Metric({
       dimensions: ifUndefined(props.dimensions, this.dimensions),
+      dimensionsMap: ifUndefined(props.dimensionsMap, props.dimensions ? props.dimensions : this.dimensions),
```

Comment:
I have to say, I don't really understand this part. `this.dimensions` is not really compatible with `props.dimensionsMap` (`dimensionsMap` has a more restrictive type), so passing `this.dimensions` for `props.dimensionsMap` seems a little weird to me.

I believe this should be just:

```suggestion
      dimensionsMap: props.dimensionsMap,
```

(If `props.dimensionsMap` is `undefined`, the code one line above that's passing `dimensions` will ensure we get `this.dimensions` to the new `Metric` instance. But definitely verify it with a unit test!)

Reply:
Ah yes, you are right! I did it the previous way because following the `ifUndefined()` pattern and this line was `dimensionsMap: ifUndefined(props.dimensionsMap, this.dimensionsMap)`, it could end up assigning `this.dimensionsMap` to the `dimensions` in the returned object when the user provided `dimensions`, but did not provide `dimensionsMap`, and the original object has some value for `dimensions`. Which I don't think would be the intent of the user if they were to call `with()` with `dimensions` and not `dimensionsMap`. But it doesn't quite work because of the type issue you called out. 

Sorry if that makes no sense! I'll be taking your change :)